oauth2client/service_account.py

```diff
@@ -12,96 +12,265 @@
+        user_agent: string, (Optional) User agent to use when sending
+                    request.
+        _token_uri: string, (Optional) The URI to use when requesting an
+                    access token.
+        _revoke_uri: string, (Optional) The URI to use when revoking an
+                     access token.
+        kwargs: dict, Extra key-value pairs to send in the payload body
+                when making an assertion.
 
-class _ServiceAccountCredentials(AssertionCredentials):
-    """Class representing a service account (signed JWT) credential."""
+        Raises:
+            ValueError: If both ``private_key_pkcs8_pem`` and
+                        ``private_key_pkcs12`` are set.
+            ValueError: If ``private_key_pkcs8_pem`` is set and
+                        ``private_key_password`` is passed in.
+            ValueError: If ``private_key_pkcs11`` is set and
+                        ``private_key_password`` is not passed in.
+    """
 
-    MAX_TOKEN_LIFETIME_SECS = 3600  # 1 hour in seconds
+    MAX_TOKEN_LIFETIME_SECS = 3600
+    """Max lifetime of the token (one hour, in seconds)."""
 
     NON_SERIALIZED_MEMBERS =  (
-        frozenset(['_private_key']) |
+        frozenset(['_signer']) |
         AssertionCredentials.NON_SERIALIZED_MEMBERS)
+    """Members that aren't serialized when object is converted to JSON."""
 
+    def __init__(self,
+                 service_account_email,
+                 private_key_pkcs8_pem=None,
+                 private_key_pkcs12=None,
+                 private_key_password=None,
+                 scopes='',
+                 private_key_id=None,
+                 service_account_id=None,
+                 user_agent=None,
+                 _token_uri=GOOGLE_TOKEN_URI,
+                 _revoke_uri=GOOGLE_REVOKE_URI,
+                 **kwargs):
 
-    def __init__(self, service_account_id, service_account_email,
-                 private_key_id, private_key_pkcs8_text, scopes,
-                 user_agent=None, token_uri=GOOGLE_TOKEN_URI,
-                 revoke_uri=GOOGLE_REVOKE_URI, **kwargs):
+        super(ServiceAccountCredentials, self).__init__(
+            None, user_agent=user_agent, token_uri=_token_uri,
+            revoke_uri=_revoke_uri)
 
-        super(_ServiceAccountCredentials, self).__init__(
-            None, user_agent=user_agent, token_uri=token_uri,
-            revoke_uri=revoke_uri)
-
-        self._service_account_id = service_account_id
         self._service_account_email = service_account_email
-        self._private_key_id = private_key_id
-        self._private_key = _get_private_key(private_key_pkcs8_text)
-        self._private_key_pkcs8_text = private_key_pkcs8_text
+        self._private_key_pkcs8_pem = private_key_pkcs8_pem
+        self._private_key_pkcs12 = private_key_pkcs12
+        self._private_key_password = private_key_password
+        self._signer = self._get_signer()
```

Comment:
It's funny I recently re-factored a research project a ton because the constructors were so hard to test/stub-out (since I used lots of instance methods in the constructor).